File: mcomix/image_handler.py
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.

from __future__ import annotations

import mimetypes
from pathlib import Path

from gi.repository import GdkPixbuf

from loguru import logger

import mcomix.image_tools as image_tools

from mcomix.file_size import format_filesize
from mcomix.image_files import ImageFiles
from mcomix.lib.events import Events, EventType
from mcomix.lib.threadpool import GlobalThreadPool, Lock
from mcomix.preferences import config
from mcomix.view_state import ViewState


class ImageHandler():
    """
    The ImageHandler keeps track of images, pages, caches and reads files.
    When the Filehandler's methods refer to pages, they are indexed from 1,
    i.e. the first page is page 1 etc.
    Other modules should *never* read directly from the files pointed to by
    paths given by the ImageHandler's methods. The files are not even
    guaranteed to exist at all times since the extraction of archives is
    threaded
    """

    def __init__(self):
        super().__init__()

        self.__events = Events()
        self.__events.add_event(EventType.FILE_AVAILABLE, self.file_available)

        self.__image_files = ImageFiles()

        #: Caching thread
        self.__threadpool = GlobalThreadPool().threadpool
        self.__lock = Lock()
        self.__cache_lock = {}
        #: Current page
        self.__current_image: int = None
        #: Set of images reading for decoding (i.e. already extracted)
        self.__available_images = set()
        #: List of pixbufs we want to cache
        self.__wanted_pixbufs: list[int] = []
        #: Pixbuf map from page > Pixbuf
        self.__raw_pixbufs: dict[int, GdkPixbuf] = {}

    @property
    def image_files(self):
        return self.__image_files

    def cleanup(self):
        # __del__ does not work for this, and there are no cpp like destructors
        self.__events.remove_event(EventType.FILE_AVAILABLE, self.file_available)

    def _get_pixbuf(self, page: int) -> GdkPixbuf:
        """
        Return the pixbuf indexed by <page> from cache.
        Pixbufs not found in cache are fetched from disk first
        """

        self._cache_pixbuf(page, force_return=False)
        return self.__raw_pixbufs[page]

    def get_pixbufs(self, number_of_bufs: int) -> list[GdkPixbuf]:
        """
        Returns number_of_bufs pixbufs for the image(s) that should be
        currently displayed. This method might fetch images from disk, so make
        sure that number_of_bufs is as small as possible
        """

        return [self._get_pixbuf(self.__current_image + i) for i in range(number_of_bufs)]

    def do_caching(self) -> None:
        """
        Make sure that the correct pixbufs are stored in cache. These
        are (in the current implementation) the current image(s), and
        if cacheing is enabled, also the one or two pixbufs before and
        after the current page. All other pixbufs are deleted and garbage
        collected directly in order to save memory
        """

        if not self.__lock.acquire():
            return

        # Get list of wanted pixbufs.
        self.__wanted_pixbufs = self._ask_for_pages(self.get_current_page())

        # remove old pixbufs.
        for page in set(self.__raw_pixbufs) - set(self.__wanted_pixbufs):
            del self.__raw_pixbufs[page]

        logger.debug(f'Caching page(s): {self.__wanted_pixbufs}')

        # Start caching available images not already in cache.
        wanted_pixbufs = [page for page in self.__wanted_pixbufs
                          if page in self.__available_images]
        self.__threadpool.map_async(self._cache_pixbuf, wanted_pixbufs)

        self.__lock.release()

    def _cache_pixbuf(self, page: int, force_return: bool = True) -> None:
        with self.__cache_lock[page]:
            if page in self.__raw_pixbufs:
                return
            with self.__lock:
                if page not in self.__wanted_pixbufs and force_return:
                    return
            logger.debug(f'Caching page: {page}')
            try:
                pixbuf = image_tools.load_pixbuf(self.__image_files.get_path_from_page(page))
            except Exception as ex:
                logger.error(f'Could not load pixbuf for page: {page}')
                logger.debug(f'Exception: {ex}')
                pixbuf = None
            self.__raw_pixbufs[page] = pixbuf

    def set_page(self, page: int) -> None:
        """
        Set up filehandler to the page <page_num>
        """

        self.__current_image = page
        self.do_caching()

    def is_page_available(self, page: int = None) -> bool:
        """
        Returns True if <page> is available and calls to get_pixbufs
        would not block. If <page> is None, the current page(s) are assumed
        """

        if page is None:
            page = self.get_current_page()

        page_list = [page]
        if ViewState.is_displaying_double and not self.is_last_page(page):
            page_list.append(page + 1)

        for page in page_list:
            if page not in self.__available_images:
                return False

        return True

    def page_available(self, page: int) -> None:
        """
        Called whenever a new page becomes available, i.e. the corresponding file has been extracted
        """

        logger.debug(f'Page is available: {page}')

        self.__cache_lock[page] = Lock()
        self.__available_images.add(page)

        # Check if we need to cache it.
        if page in self.__wanted_pixbufs:
            self.__threadpool.apply_async(self._cache_pixbuf, (page,))

        self.__events.run_events(EventType.PAGE_AVAILABLE, {'page': page})

    def file_available(self, filename: Path) -> None:
        """
        Called by the filehandler when a new file becomes available
        """

        # Find the page that corresponds to <filename>
        self.page_available(self.__image_files.get_page_from_path(filename))

    def get_number_of_pages(self) -> int:
        """
        Return the number of pages in the current archive/directory
        """

        return self.__image_files.get_total_pages()

    def get_current_page(self) -> int:
        """
        Return the current page number (starting from 1), or 0 if no file is loaded
        """

        if self.__current_image is None:
            return 0

        return self.__current_image

    def is_last_page(self, page: int = None) -> bool:
        """
        is <page> the last in a book, if page is None use current page
        """

        if page is None:
            page = self.get_current_page()

        return page == self.__image_files.get_total_pages()

    def get_path_to_page(self, page: int = None) -> Path:
        """
        Return the full path to the image file for <page>, or the current page if <page> is None
        """

        if page is None:
            page = self.__current_image

        return self.__image_files.get_path_from_page(page)

    def _get_page_unknown(self) -> str:
        if ViewState.is_displaying_double:
            return ['unknown', 'unknown']
        return ['unknown']

    def get_page_filename(self, page: int = None) -> str:
        """
        :param page
            A page number or if None the current page
        :returns
            [page, page + 1] if ViewState.displayed_double is True else return [page]
        """

        if not self.is_page_available(page=page):
            return self._get_page_unknown()

        if page is None:
            page = self.get_current_page()

        page_data = [self.get_path_to_page(page).name]

        if ViewState.is_displaying_double:
            page_data.append(self.get_path_to_page(page + 1).name)

            if ViewState.is_manga_mode:
                page_data.reverse()

        return page_data

    def get_page_filesize(self, page: int = None) -> list[str]:
        """
        :param page
            A page number or if None the current page
        :returns
            [page, page + 1] if ViewState.displayed_double is True else return [page]
        """

        if not self.is_page_available(page=page):
            return self._get_page_unknown()

        if page is None:
            page = self.get_current_page()

        page_data = [format_filesize(self.get_path_to_page(page))]

        if ViewState.is_displaying_double:
            page_data.append(format_filesize(self.get_path_to_page(page + 1)))

            if ViewState.is_manga_mode:
                page_data.reverse()

        return page_data

    def get_page_size(self, page: int = None) -> tuple[int, int]:
        """
        Return a tuple (width, height) with the size of <page>. If <page>
        is None, return the size of the current page
        """

        if page is None:
            page = self.get_current_page()

        if not self.is_page_available(page):
            return 0, 0
        pixbuf = self._get_pixbuf(page)
        return (pixbuf.get_width(), pixbuf.get_height())

    def get_mime_name(self, page: int = None) -> str:
        """
        Return a string with the name of the mime type of <page>. If
        <page> is None, return the mime type name of the current page
        """

        page_path = self.get_path_to_page(page)
        if not Path.is_file(page_path):
            return None

        mime_type, _ = mimetypes.guess_type(page_path)
        if mime_type:
            return mime_type
        return None

    def get_thumbnail(self, page: int, size: int) -> GdkPixbuf:
        """
        Return a thumbnail pixbuf of <page> that fit in a box with
        dimensions <width>x<height>. Return a thumbnail for the current
        page if <page> is None.
        If <nowait> is True, don't wait for <page> to be available
        """

        if not self._is_page_extracted(page=page):
            # Page is not available!
            return None

        path = self.get_path_to_page(page)
        return image_tools.create_thumbnail(path=path, size=size)

    def _is_page_extracted(self, page: int) -> bool:
        if page is None:
            page = self.get_current_page()

        if page in self.__available_images:
            # page is extracted
            return True

        # page is not extracted
        return False

    def _ask_for_pages(self, page: int) -> list[int]:
        """
        Ask for pages around <page> to be given priority extraction
        """

        total_pages = self.get_number_of_pages()

        cache_start = page - config['PAGE_CACHE_BEHIND']
        if cache_start < 1:
            cache_start = 1

        cache_end = page + config['PAGE_CACHE_FORWARD']
        if cache_end > total_pages:
            cache_end = total_pages

        return list(range(cache_start, cache_end))

File: src/vfs/image-handler.cxx
/**
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

#include <algorithm>
#include <filesystem>
#include <memory>
#include <ranges>
#include <string>
#include <vector>

#include <cassert>

#include "settings/settings.hxx"

#include "gui/lib/image-tools.hxx"
#include "gui/lib/view-state.hxx"

#include "vfs/image-files.hxx"
#include "vfs/image-handler.hxx"

#include "vfs/utils/utils.hxx"

#include "logger.hxx"
#include "types.hxx"

vfs::image_handler::image_handler(const std::shared_ptr<config::settings>& settings,
                                  const std::shared_ptr<gui::lib::view_state>& view_state) noexcept
    : settings(settings), view_state(view_state)
{
    image_files_ = std::make_shared<vfs::image_files>();
}

std::shared_ptr<vfs::image_files>
vfs::image_handler::image_files() const noexcept
{
    return image_files_;
}

void
vfs::image_handler::prune(const std::int32_t start, const std::int32_t size) noexcept
{
    std::erase_if(cache_,
                  [start, size](const auto& entry)
                  {
                      const auto key = entry.first;
                      return (key < start || key > (start + size));
                  });
}

#if defined(PIXBUF_BACKEND)
Glib::RefPtr<Gdk::Pixbuf>
#else
Glib::RefPtr<Gdk::Texture>
#endif
vfs::image_handler::get_image(const page_t page) noexcept
{
    if (cache_.contains(page))
    {
        // logger::trace<logger::vfs>("page {} in cache", page);
        return cache_[page];
    }

    const auto path = image_files_->path_from_page(page);

    // logger::trace<logger::vfs>("reading page {} from disk: '{}'", page, path.string());
#if defined(PIXBUF_BACKEND)
    auto image = gui::lib::image_tools::load_pixbuf(path);
#else
    auto image = gui::lib::image_tools::load_texture(path);
#endif
    if (!image)
    {
        return nullptr;
    }

    cache_.insert({page, image});
    // logger::trace<logger::vfs>("cache new size {}", cache_.size());
    return cache_[page];
}

#if defined(PIXBUF_BACKEND)
std::vector<Glib::RefPtr<Gdk::Pixbuf>>
#else
std::vector<Glib::RefPtr<Gdk::Texture>>
#endif
vfs::image_handler::get_images(const std::int32_t number) noexcept
{
    assert(number > 0);

#if defined(PIXBUF_BACKEND)
    std::vector<Glib::RefPtr<Gdk::Pixbuf>> images;
#else
    std::vector<Glib::RefPtr<Gdk::Texture>> images;
#endif
    for (const auto& i : std::views::iota(*current_image_) | std::views::take(number))
    {
        images.push_back(vfs::image_handler::get_image(i));
    }

    prune(*current_image_, static_cast<std::int32_t>(images.size()));

    return images;
}

void
vfs::image_handler::set_page(const page_t page) noexcept
{
    current_image_ = page;
}

bool
vfs::image_handler::is_page_available(const std::optional<page_t> query) const noexcept
{
    const auto page = query.value_or(get_current_page());

    std::vector<page_t> page_list = {page};
    if (view_state->is_displaying_double() && !is_last_page(page))
    {
        page_list.push_back(page + 1);
    }

    return std::ranges::all_of(page_list,
                               [this](const page_t page)
                               { return available_images_.contains(page); });
}

void
vfs::image_handler::page_available(const page_t page) noexcept
{
    // logger::debug<logger::vfs>("Page is available: {}", page);
    available_images_.insert(page);

    signal_page_available().emit(page);
}

void
vfs::image_handler::file_available(const std::filesystem::path& filename) noexcept
{
    page_available(image_files_->page_from_path(filename));
}

page_t
vfs::image_handler::get_number_of_pages() const noexcept
{
    return image_files_->total_pages();
}

page_t
vfs::image_handler::get_current_page() const noexcept
{
    return current_image_.value_or(0);
}

bool
vfs::image_handler::is_last_page(const std::optional<page_t> query) const noexcept
{
    const auto page = query.value_or(get_current_page());

    return page == image_files_->total_pages();
}

std::filesystem::path
vfs::image_handler::get_path_to_page(const std::optional<page_t> query) const noexcept
{
    const auto page = query.value_or(get_current_page());

    return image_files_->path_from_page(page);
}

const std::vector<std::string>
vfs::image_handler::get_page_filename(const std::optional<page_t> query) const noexcept
{
    const auto page = query.value_or(get_current_page());

    std::vector<std::string> page_data;

    if (!is_page_available(page))
    {
        page_data.push_back("unknown");
        if (view_state->is_displaying_double())
        {
            page_data.push_back("unknown");
        }
        return page_data;
    }

    page_data.push_back(get_path_to_page(page));

    if (view_state->is_displaying_double())
    {
        page_data.push_back(get_path_to_page(page + 1));

        if (view_state->is_manga_mode())
        {
            std::ranges::reverse(page_data);
        }
    }
    return page_data;
}

const std::vector<std::string>
vfs::image_handler::get_page_filesize(const std::optional<page_t> query) const noexcept
{
    const auto page = query.value_or(get_current_page());

    std::vector<std::string> page_data;

    if (!is_page_available(page))
    {
        page_data.push_back("unknown");
        if (view_state->is_displaying_double())
        {
            page_data.push_back("unknown");
        }
        return page_data;
    }

    page_data.push_back(vfs::utils::file_size(get_path_to_page(page), settings->si_units));

    if (view_state->is_displaying_double())
    {
        page_data.push_back(vfs::utils::file_size(get_path_to_page(page + 1), settings->si_units));

        if (view_state->is_manga_mode())
        {
            std::ranges::reverse(page_data);
        }
    }
    return page_data;
}

std::array<std::int32_t, 2>
vfs::image_handler::get_page_size(const std::optional<page_t> query) noexcept
{
    const auto page = query.value_or(get_current_page());

    std::array<std::int32_t, 2> dimensions = {0, 0};

    if (!is_page_available(page))
    {
        return dimensions;
    }

    auto image = get_image(page);
    if (!image)
    {
        return dimensions;
    }

    return {image->get_width(), image->get_height()};
}

std::string
vfs::image_handler::get_mime_name(const std::optional<page_t> query) const noexcept
{
    // TODO
    const auto page = query.value_or(get_current_page());
    (void)page;
    return "";
}

Glib::RefPtr<Gdk::Paintable>
vfs::image_handler::get_thumbnail(const page_t page, const std::int32_t size) noexcept
{
    if (!is_page_extracted(page))
    {
        return nullptr;
    }

    return gui::lib::image_tools::create_thumbnail(get_image(page), size);
}

bool
vfs::image_handler::is_page_extracted(const std::optional<page_t> query) const noexcept
{
    const auto page = query.value_or(get_current_page());

    return std::ranges::contains(available_images_, page);
}
